File: src/com/wang/Game2dEngine/monitor/InputMonitor.java
package com.wang.Game2dEngine.monitor;

import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.MouseMotionListener;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
/**
 *
 * @author ricolwang
 */
public class InputMonitor implements MouseListener, KeyListener, MouseMotionListener
{

    public interface IKeyTyped
    {

        void inputMonitorKeyTyped(char key);
    }

    public boolean leftButtonPressed = false;
    public boolean rightButtonPressed = false;
    public boolean middleButtonPressed = false;
    public int MouseX;
    public int MouseY;
    public boolean mouseEntered = false;

    static InputMonitor instance = new InputMonitor();
    HashMap<Integer, Boolean> keyStatus = new HashMap<>();
    Set<IKeyTyped> keytypedobserver = new HashSet<>();

    public static InputMonitor getSharedInstance()
    {
        return instance;
    }

    @Override
    public void mouseClicked(MouseEvent e)
    {

    }

    @Override
    public void mousePressed(MouseEvent e)
    {
        switch (e.getButton())
        {
            case MouseEvent.BUTTON1:
                leftButtonPressed = true;
                break;
            case MouseEvent.BUTTON2:
                middleButtonPressed = true;
                break;
            case MouseEvent.BUTTON3:
                rightButtonPressed = true;
                break;
            default:
                break;
        }
    }

    @Override
    public void mouseReleased(MouseEvent e)
    {
        switch (e.getButton())
        {
            case MouseEvent.BUTTON1:
                leftButtonPressed = false;
                break;
            case MouseEvent.BUTTON2:
                middleButtonPressed = false;
                break;
            case MouseEvent.BUTTON3:
                rightButtonPressed = false;
                break;
            default:
                break;
        }
    }

    @Override
    public void mouseEntered(MouseEvent e)
    {
        mouseEntered = true;
    }

    @Override
    public void mouseExited(MouseEvent e)
    {
        mouseEntered = false;
    }

    @Override
    public void keyTyped(KeyEvent e)
    {
        for (IKeyTyped i : keytypedobserver)
        {
            i.inputMonitorKeyTyped(e.getKeyChar());
        }
    }

    @Override
    public void keyPressed(KeyEvent e)
    {
        this.updateKeyStatus(e.getKeyCode(), true);
    }

    @Override
    public void keyReleased(KeyEvent e)
    {
        this.updateKeyStatus(e.getKeyCode(), false);
    }

    @Override
    public void mouseDragged(MouseEvent e)
    {

    }

    @Override
    public void mouseMoved(MouseEvent e)
    {
        MouseX = e.getX();
        MouseY = e.getY();
    }

    public boolean isKeyPressed(int key)
    {
        return keyStatus.getOrDefault(key, false);
    }

    void updateKeyStatus(int key, boolean status)
    {
        keyStatus.put(key, status);
    }

    public void addObserverForKeyTyped(IKeyTyped o)
    {
        this.keytypedobserver.add(o);
    }

    public void removeObserverForKeyTyped(IKeyTyped o)
    {
        this.keytypedobserver.remove(o);
    }
}

File: src/com/wang/Game2dEngine/monitor/InputMonitor.kt
package com.wang.Game2dEngine.monitor

import java.awt.event.*
import java.util.*

/*
* To change this license header, choose License Headers in Project Properties.
* To change this template file, choose Tools | Templates
* and open the template in the editor.
*/
/**
 * @author ricolwang
 */
class InputMonitor : MouseListener, KeyListener, MouseMotionListener
{
    interface IKeyTyped
    {
        fun inputMonitorKeyTyped(key: Char)
    }

    var leftButtonPressed = false
    var rightButtonPressed = false
    var middleButtonPressed = false
    var MouseX = 0
    var MouseY = 0
    var mouseEntered = false
    var keyStatus = HashMap<Int, Boolean>()
    var keytypedobserver: MutableSet<IKeyTyped> = HashSet()
    override fun mouseClicked(e: MouseEvent)
    {
    }

    override fun mousePressed(e: MouseEvent)
    {
        when (e.button)
        {
            MouseEvent.BUTTON1 -> leftButtonPressed = true
            MouseEvent.BUTTON2 -> middleButtonPressed = true
            MouseEvent.BUTTON3 -> rightButtonPressed = true
            else ->
            {
            }
        }
    }

    override fun mouseReleased(e: MouseEvent)
    {
        when (e.button)
        {
            MouseEvent.BUTTON1 -> leftButtonPressed = false
            MouseEvent.BUTTON2 -> middleButtonPressed = false
            MouseEvent.BUTTON3 -> rightButtonPressed = false
            else ->
            {
            }
        }
    }

    override fun mouseEntered(e: MouseEvent)
    {
        mouseEntered = true
    }

    override fun mouseExited(e: MouseEvent)
    {
        mouseEntered = false
    }

    override fun keyTyped(e: KeyEvent)
    {
        for (i in keytypedobserver)
        {
            i.inputMonitorKeyTyped(e.keyChar)
        }
    }

    override fun keyPressed(e: KeyEvent)
    {
        updateKeyStatus(e.keyCode, true)
    }

    override fun keyReleased(e: KeyEvent)
    {
        updateKeyStatus(e.keyCode, false)
    }

    override fun mouseDragged(e: MouseEvent)
    {
    }

    override fun mouseMoved(e: MouseEvent)
    {
        MouseX = e.x
        MouseY = e.y
    }

    fun isKeyPressed(key: Int): Boolean
    {
        return keyStatus.getOrDefault(key, false)
    }

    fun updateKeyStatus(key: Int, status: Boolean)
    {
        keyStatus[key] = status
    }

    fun addObserverForKeyTyped(o: IKeyTyped)
    {
        keytypedobserver.add(o)
    }

    fun removeObserverForKeyTyped(o: IKeyTyped)
    {
        keytypedobserver.remove(o)
    }

    companion object
    {
        var sharedInstance = InputMonitor()
    }
}